config/database.go

```diff
@@ -0,0 +1,322 @@
+
+	// TODO: disallow attempting to save a directly modified config (comparing pointers). This
+	// wouldn't be an exhaustive check, given the use of pointers throughout the data
+	// structure, but might prevent common mistakes. Requires upstream changes first.
+	// if newCfg == oldCfg {
+	// 	return nil, errors.New("old configuration modified instead of cloning")
+	// }
+
+	newCfg = newCfg.Clone()
+	newCfg.SetDefaults()
+
+	// Sometimes the config is received with "fake" data in sensitive fields. Apply the real
+	// data from the existing config as necessary.
+	desanitize(oldCfg, newCfg)
+
+	if err := newCfg.IsValid(); err != nil {
+		return nil, errors.Wrap(err, "new configuration is invalid")
+	}
+
+	// Ideally, Set would persist automatically and abstract this completely away from the
+	// client. Doing so requires a few upstream changes first, so for now an explicit Save()
+	// remains required.
+	// if err := ds.persist(newCfg); err != nil {
+	// 	return nil, errors.Wrap(err, "failed to persist")
+	// }
+
+	ds.config = newCfg
+
+	unlockOnce.Do(ds.configLock.Unlock)
+
+	// Notify listeners synchronously. Ideally, this would be asynchronous, but existing code
+	// assumes this and there would be increased complexity to avoid racing updates.
+	ds.invokeConfigListeners(oldCfg, newCfg)
+
+	return oldCfg, nil
+}
+
+// persist writes the configuration to the configured database.
+func (ds *DatabaseStore) persist(cfg *model.Config) error {
+	b, err := marshalConfig(cfg)
+	if err != nil {
+		return errors.Wrap(err, "failed to serialize")
+	}
+
+	id := model.NewId()
+	value := string(b)
+	createAt := model.GetMillis()
+
+	tx, err := ds.db.Beginx()
+	if err != nil {
+		return errors.Wrap(err, "failed to begin transaction")
+	}
+	defer func() {
+		// Rollback after Commit just returns sql.ErrTxDone.
+		if err := tx.Rollback(); err != nil && err != sql.ErrTxDone {
+			mlog.Error("Failed to rollback configuration transaction", mlog.Err(err))
+		}
+	}()
+
+	params := map[string]interface{}{
+		"id":        id,
+		"value":     value,
+		"create_at": createAt,
+		"key":       "ConfigurationId",
+	}
+
+	if _, err := tx.Exec("UPDATE Configurations SET Active = NULL WHERE Active"); err != nil {
+		return errors.Wrap(err, "failed to deactivate current configuration")
+	}
+
+	if _, err := tx.NamedExec("INSERT INTO Configurations (Id, Value, CreateAt, Active) VALUES (:id, :value, :create_at, TRUE)", params); err != nil {
+		return errors.Wrap(err, "failed to record new configuration")
+	}
+
+	if err := tx.Commit(); err != nil {
+		return errors.Wrap(err, "failed to commit transaction")
+	}
+
+	return nil
+}
+
+// initializeConfigurationsTable ensures the requisite tables in place to form the backing store.
+func initializeConfigurationsTable(db *sqlx.DB) error {
+	_, err := db.Exec(`
+		CREATE TABLE IF NOT EXISTS Configurations (
+		    Id VARCHAR(26) PRIMARY KEY,
+		    Value TEXT NOT NULL,
+		    CreateAt BIGINT NOT NULL,
+		    Active BOOLEAN NULL UNIQUE
```

Comment:
Why `Active` instead of use the `CreateAt`?

Reply:
Leveraging `CreateAt` exposes a race condition: if two servers each try to update the config, the latter one will silently replace the earlier version without a chance to reconcile the changes. By converging on `Active`, one of the transactions will be forced to roll back and can decide how it wants to handle the situation.

/Today/, it doesn't much matter, since we tend to happily clobber the config even if it might have changed, but this becomes somewhat more interesting when we support a `Patch` operation.

Reply:
Good point :+1: